lib/sdf-server/src/garbage_collection.rs

```diff
@@ -0,0 +1,156 @@
+        );
+        let change_set_snapshot_rows = ctx.txns().await?.pg().query(
+            "SELECT workspace_snapshot_address AS snapshot_id FROM change_set_pointers WHERE status = ANY($1::text[]) GROUP BY workspace_snapshot_address",
+            &[&open_statuses],
+        ).await?;
+        for row in change_set_snapshot_rows {
+            let snapshot_id: WorkspaceSnapshotAddress = row.try_get("snapshot_id")?;
+            open_change_set_snapshot_ids.insert(snapshot_id);
+        }
+        info!(
+            "Found {} distinct snapshot address(es) for open change sets.",
+            open_change_set_snapshot_ids.len()
+        );
+
+        // Gather the WorkspaceSnapshotAddress of all existing snapshots that are
+        // at least an hour old.
+        //
+        // By only considering the snapshots that are at least an hour old, we avoid
+        // race conditions where a change set is created, or modified between when we
+        // queried the change_set_pointers table and the workspace_snapshots table.
+        // We can't rely on transactional integrity to avoid race conditions as the
+        // tables are in completely separate databases.
+        let snapshot_id_rows = ctx
+            .layer_db()
+            .workspace_snapshot()
+            .cache
+            .pg()
+            .query(
+                "SELECT key AS snapshot_id FROM workspace_snapshots WHERE created_at < NOW() - '1 hour'::interval GROUP BY key",
+                &[],
+            )
+            .await?
+            .ok_or_else(|| GarbageCollectorError::UnableToQuerySnapshots)?;
+        for row in snapshot_id_rows {
+            let snapshot_id: WorkspaceSnapshotAddress = row.try_get("snapshot_id")?;
+            all_snapshot_ids.insert(snapshot_id);
+        }
+        info!(
+            "Found {} distinct snapshot address(es) older than cutoff.",
+            all_snapshot_ids.len()
+        );
+
+        // Any WorkspaceSnapshotAddress not in both open_change_set_snapshot_ids
+        // and all_snapshot_ids is for a closed/applied/abandoned change set and
+        // can be deleted, as we do not allow re-opening change sets.
+        let snapshot_ids_to_delete: HashSet<_> = all_snapshot_ids
+            .difference(&open_change_set_snapshot_ids)
+            .collect();
+        info!(
+            "Found {} snapshot address(es) only used by inactive change sets.",
+            snapshot_ids_to_delete.len()
+        );
+
+        let mut counter = 0;
+        for key in snapshot_ids_to_delete {
+            ctx.layer_db()
+                .workspace_snapshot()
+                .cache
+                .pg()
+                .delete(&key.to_string())
+                .await?;
+
+            counter += 1;
+            if counter % 100 == 0 {
+                info!("Deleted {} snapshot addresses.", counter);
+            }
+        }
+        info!("Deleted {} snapshot address(es).", counter);
+
+        ctx.commit().await?;
```

Comment:
This shouldn't have any interaction with the layer cache's database, only for the DAL db and any pending WS events.